src/main/cc/wfa/virtual_people/core/model/branch_node_impl_test.cc

```diff
@@ -0,0 +1,362 @@
+      branch_node {
+        branches {
+          node_index: 2
+          chance: 0.4
+        }
+        branches {
+          node_index: 3
+          chance: 0.6
+        }
+        random_seed: "TestBranchNodeSeed"
+      }
+  )pb", &branch_node_config));
+  ASSERT_OK_AND_ASSIGN(
+      std::unique_ptr<ModelNode> branch_node,
+      ModelNodeFactory().NewModelNode(branch_node_config));
+
+  LabelerEvent input;
+  input.set_acting_fingerprint(0);
+  EXPECT_THAT(
+    branch_node->Apply(input),
+    StatusIs(absl::StatusCode::kFailedPrecondition, ""));
+}
+
+TEST(BranchNodeImplTest, TestNoBranch) {
+  CompiledNode config;
+  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(R"pb(
+      name: "TestBranchNode"
+      index: 1
+      branch_node {}
+  )pb", &config));
+  EXPECT_THAT(
+      ModelNodeFactory().NewModelNode(config).status(),
+      StatusIs(absl::StatusCode::kInvalidArgument, ""));
+}
+
+TEST(BranchNodeImplTest, TestNoChildNode) {
+  CompiledNode config;
+  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(R"pb(
+      name: "TestBranchNode"
+      index: 1
+      branch_node {
+        branches {
+          chance: 0.5
+        }
+      }
+  )pb", &config));
+  EXPECT_THAT(
+      ModelNodeFactory().NewModelNode(config).status(),
+      StatusIs(absl::StatusCode::kInvalidArgument, ""));
+}
+
+TEST(BranchNodeImplTest, TestNoSelectBy) {
+  CompiledNode config;
+  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(R"pb(
+      name: "TestBranchNode"
+      index: 1
+      branch_node {
+        branches {
+          node_index: 2
+        }
+      }
+  )pb", &config));
+  EXPECT_THAT(
+      ModelNodeFactory().NewModelNode(config).status(),
+      StatusIs(absl::StatusCode::kInvalidArgument, ""));
+}
+
+TEST(BranchNodeImplTest, TestDifferentSelectBy) {
+  CompiledNode config;
+  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(R"pb(
+      name: "TestBranchNode"
+      index: 1
+      branch_node {
+        branches {
+          node_index: 2
+          chance: 0.5
+        }
+        branches {
+          node_index: 3
+          condition {}
+        }
+      }
+  )pb", &config));
+  EXPECT_THAT(
+      ModelNodeFactory().NewModelNode(config).status(),
+      StatusIs(absl::StatusCode::kInvalidArgument, ""));
+}
+
+TEST(BranchNodeImplTest, TestResolveChildReferencesIndexNotFound) {
+  CompiledNode config;
+  ASSERT_TRUE(google::protobuf::TextFormat::ParseFromString(R"pb(
+      name: "TestBranchNode"
+      index: 1
+      branch_node {
+        branches {
+          node_index: 2
+          chance: 1
+        }
+        random_seed: "TestBranchNodeSeed"
+      }
+  )pb", &config));
+  ASSERT_OK_AND_ASSIGN(
+      std::unique_ptr<ModelNode> node,
+      ModelNodeFactory().NewModelNode(config));
+
+  absl::flat_hash_map<uint32_t, std::unique_ptr<ModelNode>> node_refs;
+  EXPECT_THAT(
+      node->ResolveChildReferences(node_refs),
+      StatusIs(absl::StatusCode::kInvalidArgument, ""));
+}
```

Comment:
Add a test case for nested child reference, i.e. a node_index child also uses node_index.